### PR TITLE
Make Sneak Peek respect release dates

### PR DESCRIPTION
**Pull request by caesar2164**
Make Sneak Peek/Direct Access respect release dates of content.

### CODE REVIEW

**Comment by caesar2164** [comment]
@stvstnfrd 

Course Outline:
screen shot 2016-01-27 at 2 53 00 pm

What Staff Sees:
screen shot 2016-01-27 at 2 52 24 pm

What Direct Access Sees:
screen shot 2016-01-27 at 2 51 47 pm

**Comment by dcadams** [comment]
Can we have a unit test/s such that the previous code fails but with the fix it passes?

**Comment by stvstnfrd** [comment]
> PS: styling is weird due to me running paver devstack studio AFTER paver destack lms...

I'm not familiar with this; what's that?

File: lms/djangoapps/courseware/access.py [hunk 1]
```diff
@@ -618,13 +618,18 @@ def can_load():
...
                 if _can_load_descriptor_nonregistered(descriptor):
-                    return ACCESS_GRANTED
+                    sneakpeek_access = ACCESS_GRANTED
         response = (
             _visible_to_nonstaff_users(descriptor)
-            and _has_group_access(descriptor, user, course_key)
+            and
+            (
+                _has_group_access(descriptor, user, course_key)
+                or sneakpeek_access
```
Comment (stvstnfrd) [inline lms/djangoapps/courseware/access.py]: > section 3 is hidden/staff only

I'm still seeing Section 3 in the screenshot.

I don't think this is correct. Remember, Direct Access users really are enrolled users, with an entry in `auth_user`.
What this expression is saying is:

> If their group deserves access, let them in.
> Even if their group doesn't have access, go ahead and let them in anyway, if they're sneak_peek.

So I suspect that sneak_peek users will mistakenly slide through if they're in a "normal group" that doesn't have access.

I don't think `_has_group_access` and `sneakpeek_access` are logically connected, so I'm not sure that grouping them together with an `or` is appropriate.

If you look at `_has_group_access` [1], any check that can `return ACCESS_DENIED` will be ignored for sneak_peek users. This one in particular jumps out [2]:
```python
log.warning("Group access check excludes all students, access will be denied.", exc_info=True)
```

[1] https://github.com/Stanford-Online/edx-platform/blob/giulio/sneakpeek-release-date-bugfix/lms/djangoapps/courseware/access.py#L537
[2] https://github.com/Stanford-Online/edx-platform/blob/giulio/sneakpeek-release-date-bugfix/lms/djangoapps/courseware/access.py#L554

**Comment by stvstnfrd** [comment]
See my concerns above.

The more I look at this method, the more I think you just can't reasonably inject a short-circuit for sneak_peak directly into this boolean expression; it's already wonky as-is.

Really, this method should be written in the form:
```python
# Let staff in
if _has_staff_access_to_descriptor(user, descriptor, course_key):
    return ACCESS_GRANTED
# Block staff-only
if not _visible_to_nonstaff_users(descriptor):
    return ACCESS_DENIED
# Block sneak_peak users for non-whitelisted categories
if user.is_authenticated():
    if not UserProfile.has_registered(user):
        if not _can_load_descriptor_nonregistered(descriptor):
            return ACCESS_DENIED
# Check group access
if not _has_group_access(descriptor, user, course_key):
    return ACCESS_DENIED
# Enforce start_date
if not _can_access_descriptor_with_start_date(user, descriptor, course_key) and not _has_detached_class_tag(descriptor):
    return ACCESS_DENIED
# Otherwise, let them in
return ACCESS_GRANTED
```

So basically, let staff in no matter what. Otherwise, go through permission by permission, and DENY if they don't pass.

Not only is this simpler to follow, but every check _after_ staff access can appear in any order.

While that's ideal, I don't think we want to make this drastic of a fork right now. This is why I took the approach I did in #420. Essentially what it does is say:

> If they're sneak_peak but sneak_peak users aren't allowed to view this type of descriptor, go ahead and DENY. Otherwise, check permissions how you normally would.

On a high level, that sounds like the check we want to make.

My branch does have test failures, but I'm not sure that's incorrect. It looks like most of those are discussion failures where it appears to be expecting sneak_peak users to have access to parts of the forum when in reality, they shouldn't have access anyway. @dcadams This looks eerily similar to #386

File: lms/djangoapps/courseware/access.py (hunk 1 above)
Comment (caesar2164) [inline lms/djangoapps/courseware/access.py]: > I'm still seeing Section 3 in the screenshot.

By section 3 I mean the 3rd section which is called: **Example Week 2: Get Interactive**

DIrect Access users **do** have an account, but they haven't registered so it fails `_has_group_access`

**Comment by caesar2164** [comment]
I understand what your way does, but my contention is that DA users would fail the `_has_group_access ` check, and get booted...

**Comment by caesar2164** [comment]
ok, @nthaniel & I discussed DA + cohorts/groups and we came to the conclusion that DA users should **NOT** see cohorted content. So I will close this PR in favor of yours, @stvstnfrd.

**Pull request closed without merging**